cocos/core/game.ts

```diff
@@ -610,214 +462,164 @@ export class Game extends EventTarget {
-        // we allow the `run(config, onstart?)` form. But it's deprecated.
-        let initPromise: Promise<boolean> | undefined;
-        if (typeof configOrCallback !== 'function' && configOrCallback) {
-            initPromise = this.init(configOrCallback);
-            this.onStart = onStart ?? null;
-        } else {
-            this.onStart = configOrCallback ?? null;
-        }
-        garbageCollectionManager.init();
-
-        return Promise.resolve(initPromise)
-            .then(() => Promise.resolve(builtinResMgr.initBuiltinRes(this._gfxDevice!)))
-            .then(() => this._setRenderPipelineNShowSplash()).then(() => {
-                if (!HTML5) {
-                    // @ts-expect-error access private method.
-                    screen._init({
-                        configOrientation: 'auto',
-                        exactFitScreen: true,
+            })
+            .then(() => {
+                this.emit(Game.EVENT_POST_BASE_INIT);
+                return this.onPostBaseInitDelegate.dispatch();
+            })
+            // #endregion Base
+            // #region Settings
+            .then(() => {
+                this.emit(Game.EVENT_PRE_SETTINGS_INIT);
+                return this.onPreSettingsInitDelegate.dispatch();
+            })
+            .then(() => settings.init(config.settingsPath, config.overrideSettings))
+            .then(() => {
+                this.emit(Game.EVENT_POST_SETTINGS_INIT);
+                return this.onPostSettingsInitDelegate.dispatch(settings);
+            })
+            // #endregion Settings
+            // #region Infrastructure
+            .then(() => {
+                this.emit(Game.EVENT_PRE_INFRASTRUCTURE_INIT);
+                return this.onPreInfrastructureInitDelegate.dispatch();
+            })
+            .then(() => {
+                macro.init();
+                const adapter = findCanvas();
+                if (adapter) {
+                    this.canvas = adapter.canvas;
+                    this.frame = adapter.frame;
+                    this.container = adapter.container;
+                }
+                screen.init();
+                garbageCollectionManager.init();
+                deviceManager.init(this.canvas, bindingMappingInfo);
+                assetManager.init();
+                builtinResMgr.init();
+                Layers.init();
+                this.initPacer();
+            })
+            .then(() => {
+                this.emit(Game.EVENT_POST_INFRASTRUCTURE_INIT);
+                return this.onPostInfrastructureInitDelegate.dispatch();
+            })
+            // #endregion Infrastructure
+            // #region Subsystem
+            .then(() => {
+                this.emit(Game.EVENT_PRE_SUBSYSTEM_INIT);
+                return this.onPreSubsystemInitDelegate.dispatch();
+            })
+            .then(() => director.init())
+            .then(() => builtinResMgr.loadBuiltinAssets())
```

Comment:
should move to infrastructure region, However, it is limited by the fact that materials must currently rely on the renderer, so delaying until after the subsystem is initialized